### PR TITLE
feat: add host plugins

### PR DESCRIPTION
**Pull request by junyu0312**

### CODE REVIEW

File: crates/playground/examples/binary_search.rs [hunk 1]
```diff
@@ -2,14 +2,15 @@ use std::cell::RefCell;
 use std::rc::Rc;
 
 use anyhow::Result;
-use delphinus_zkwasm::loader::ExecutionArg;
 use delphinus_zkwasm::loader::ZkWasmLoader;
+use delphinus_zkwasm::runtime::host::default_env::DefaultHostEnvBuilder;
```
Comment (junyu0312) [inline crates/playground/examples/binary_search.rs:6]: keep playground unchanged

File: crates/zkwasm/src/test/test_rlp.rs [hunk 2]
```diff
@@ -2,14 +2,19 @@ use std::cell::RefCell;
 use std::rc::Rc;
 
 use crate::circuits::TestCircuit;
-use crate::loader::ExecutionArg;
 use crate::loader::ZkWasmLoader;
+use crate::runtime::host::default_env::DefaultHostEnvBuilder;
```
Comment (junyu0312) [inline crates/zkwasm/src/test/test_rlp.rs:6]: revert

Comment (junyu0312) [inline crates/zkwasm/src/test/test_rlp.rs:6]: revert

File: crates/zkwasm/Cargo.toml [hunk 3]
```diff
@@ -22,12 +22,26 @@ strum = "0.24.1"
 strum_macros = "0.24.1"
 serde = { version = "1.0", features = ["derive"] }
 serde_json = "1.0"
+
+ff = "0.12"
+sha2 = "0.10.6"
+poseidon = { git = "https://github.com/lanbones/poseidon" }
+#parity-wasm = { version = "0.42.0", default-features = false }
```
Comment (junyu0312) [inline crates/zkwasm/Cargo.toml:29]: remove comments

File: README.md [hunk 4]
```diff
@@ -31,7 +31,7 @@ clang lld
 wasm code
 
 ## Runtime input:
-input of wasm function and the top level function must be main.
+input of wasm function and the name of the top level function (usually we call the entry function zkmain).
```
Comment (junyu0312) [inline README.md:34]: must be `zkmain`

File: crates/cli/src/exec.rs [hunk 5]
```diff
@@ -30,30 +23,30 @@ use serde::Deserialize;
...
 use wasmi::RuntimeValue;
+use std::io::Write;
+use delphinus_host::StandardHostEnvBuilder;
+use delphinus_host::ExecutionArg;
 
-use crate::app_builder::write_context_output;
-use crate::args::parse_args;
-
-const AGGREGATE_PREFIX: &'static str = "aggregate-circuit";
+//const AGGREGATE_PREFIX: &'static str = "aggregate-circuit";
```
Comment (junyu0312) [inline crates/cli/src/exec.rs:34]: remove it

File: crates/cli/src/exec.rs [hunk 6]
```diff
@@ -70,13 +63,17 @@ pub fn exec_setup(
...
     {
-        let vk_path = &output_dir.join(format!("{}.{}.vkey.data", prefix, 0));
+        let vk_path = &param_dir.join(format!("{}.vkey.data", prefix));
 
         if vk_path.exists() {
             info!("Found Verifying at {:?}", vk_path);
         } else {
             info!("Create Verifying to {:?}", vk_path);
-            let loader = ZkWasmLoader::<Bn256>::new(zkwasm_k, wasm_binary, phantom_functions)?;
+            let loader = ZkWasmLoader::<Bn256, ExecutionArg, StandardHostEnvBuilder>::new(
```
Comment (junyu0312) [inline crates/cli/src/exec.rs:72]: 'DefaultHostEnvBuilder' from `app_builder.rs` and 'StandardHostEnvBuilder' should be unified.

Comment (junyu0312) [reply crates/cli/src/exec.rs:72]: Alternatively, you might use generic type.

File: crates/cli/src/exec.rs [hunk 7]
```diff
@@ -94,7 +91,11 @@ pub fn exec_image_checksum(
     phantom_functions: Vec<String>,
     output_dir: &PathBuf,
 ) -> Result<()> {
-    let loader = ZkWasmLoader::<Bn256>::new(zkwasm_k, wasm_binary, phantom_functions)?;
+    let loader = ZkWasmLoader::<Bn256, ExecutionArg, StandardHostEnvBuilder>::new(
```
Comment (junyu0312) [inline crates/cli/src/exec.rs:94]: the same

File: crates/cli/src/exec.rs [hunk 8]
```diff
@@ -172,19 +173,20 @@ pub fn exec_dry_run_service(
...
 
-                            let loader = ZkWasmLoader::<Bn256>::new(
-                                zkwasm_k,
-                                wasm_binary.clone(),
-                                phantom_functions.clone(),
-                            )
-                            .unwrap();
-
+                            let loader =
+                                ZkWasmLoader::<Bn256, ExecutionArg, StandardHostEnvBuilder>::new(
```
Comment (junyu0312) [inline crates/cli/src/exec.rs:177]: the same

File: crates/cli/src/app_builder.rs [hunk 9]
```diff
@@ -162,99 +165,35 @@ pub trait AppBuilder: CommandBuilder {
...
                 )?;
 
                 write_context_output(&context_out.borrow(), context_out_path)?;
 
                 Ok(())
             }
             Some(("single-verify", sub_matches)) => {
-                let proof_path: PathBuf = Self::parse_proof_path_arg(&sub_matches);
-                let instance_path: PathBuf = Self::parse_single_instance_arg(&sub_matches);
+                let _proof_path: PathBuf = Self::parse_proof_path_arg(&sub_matches);
```
Comment (junyu0312) [inline crates/cli/src/app_builder.rs:188]: remove them if they are no longer useful, as well as the related code in 'append_verify_single_proof_subcommand'.

File: crates/cli/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,4 @@
+pub mod app_builder;
```
Comment (junyu0312) [inline crates/cli/src/lib.rs:1]: cli is also used as a lib?

File: crates/cli/test_cli.sh [hunk 11]
```diff
@@ -6,14 +6,8 @@ set -x
...
-
-RUST_LOG=info cargo run --release --features cuda -- -k 18 --function zkmain --output ./output --wasm ../zkwasm/wasm/wasm_output.wasm aggregate-prove --public 133:i64 --public 2:i64
-RUST_LOG=info cargo run --release --features cuda -- -k 18 --function zkmain --output ./output --wasm ../zkwasm/wasm/wasm_output.wasm aggregate-verify --proof output/aggregate-circuit.0.transcript.data  --instances output/aggregate-circuit.0.instance.data
-if [ -d "sol" ]; then
-  RUST_LOG=info cargo run --release --features cuda -- -k 18 --function zkmain --output ./output --wasm ../zkwasm/wasm/wasm_output.wasm solidity-aggregate-verifier --proof output/aggregate-circuit.0.transcript.data  --instances output/aggregate-circuit.0.instance.data
-fi
+RUST_LOG=info cargo run --release --features cuda -- -k 18 --function zkmain --param ./params --output ./output --wasm ../zkwasm/wasm/wasm_output.wasm single-prove --public 133:i64 --public 2:i64
+RUST_LOG=info cargo run --release --features cuda -- -k 18 --function zkmain --param ./params --output ./output --wasm ../zkwasm/wasm/wasm_output.wasm single-verify --proof output/zkwasm.0.transcript.data --instance output/zkwasm.0.instance.data
```
Comment (junyu0312) [inline crates/cli/test_cli.sh:13]: remove `proof` and `instance` since they are unused.

File: crates/zkwasm/src/circuits/mod.rs [hunk 12]
```diff
@@ -36,6 +39,13 @@ impl<F: FieldExt> TestCircuit<F> {
             _data: PhantomData,
         }
     }
+
+    pub fn new_without_configure(tables: Tables) -> Self {
```
Comment (junyu0312) [inline crates/zkwasm/src/circuits/mod.rs:43]: the function can be deleted now

File: test_bls.sh [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+set -e
+set -x
+
+# rm -rf output
+
+# Single test
+RUST_LOG=info cargo run --release  -- -k 20 --function zkmain --output ./output --wasm ctest/bls.wasm setup
+
+RUST_LOG=info cargo run --release  -- -k 20 --function zkmain --output ./output --wasm ctest/bls.wasm single-prove
+RUST_LOG=info cargo run --release  -- -k 20 --function zkmain --output ./output --wasm ctest/bls.wasm single-verify --proof output/zkwasm.0.transcript.data
```
Comment (junyu0312) [inline test_bls.sh:12]: remove `proof`, remove `aggregate-prove` and `aggregate-verify`

File: crates/cli/src/lib.rs (hunk 10 above)
Comment (xgaozoyoe) [reply crates/cli/src/lib.rs:1]: cli is only used by this repo. The wasm playground is connected straight to zkwasm lib.

File: test_bls.sh (hunk 13 above)
Comment (xgaozoyoe) [reply test_bls.sh:12]: done

File: crates/zkwasm/src/circuits/mod.rs (hunk 12 above)
Comment (xgaozoyoe) [reply crates/zkwasm/src/circuits/mod.rs:43]: done

File: crates/cli/src/lib.rs (hunk 10 above)
Comment (junyu0312) [reply crates/cli/src/lib.rs:1]: So why add the lib.rs file

File: crates/zkwasm/Cargo.toml [hunk 14]
```diff
@@ -22,12 +22,21 @@ strum = "0.24.1"
 strum_macros = "0.24.1"
 serde = { version = "1.0", features = ["derive"] }
 serde_json = "1.0"
+
```
Comment (junyu0312) [inline crates/zkwasm/Cargo.toml:25]: remove blank lines

File: crates/cli/src/exec.rs [hunk 15]
```diff
@@ -172,13 +171,13 @@ pub fn exec_dry_run_service(
...
 
-                            let loader = ZkWasmLoader::<Bn256>::new(
-                                zkwasm_k,
-                                wasm_binary.clone(),
-                                phantom_functions.clone(),
-                            )
-                            .unwrap();
-
+                            let loader =
+                                ZkWasmLoader::<Bn256, ExecutionArg, DefaultHostEnvBuilder>::new(
```
Comment (junyu0312) [inline crates/cli/src/exec.rs:175]: replace `ExecutionArg`, `DefaultHostEnvBuilder` with `Arg`, `Builder`

File: test_bls.sh (hunk 13 above)
Comment (junyu0312) [reply test_bls.sh:12]: You haven't deleted them

File: crates/cli/src/lib.rs (hunk 10 above)
Comment (xgaozoyoe) [reply crates/cli/src/lib.rs:1]: correct my last comment. It is also used as a library by some customers.

**Review by xgaozoyoe — APPROVED** [review]

**Review by lanbones — APPROVED** [review]

**Pull request merged**